pubsub/subscription.go

```diff
@@ -1372,9 +1388,32 @@ func (s *Subscription) Receive(ctx context.Context, f func(context.Context, *Mes
 					return nil
 				default:
 				}
+
 				for i, msg := range msgs {
 					msg := msg
-					// TODO(jba): call acquire closer to when the message is allocated.
+					iter.eoMu.RLock()
+					ackh, _ := msgAckHandler(msg, iter.enableExactlyOnceDelivery)
+					iter.eoMu.RUnlock()
+					// otelCtx is used to relate the main subscribe span to the other subspans in the subscribe side.
+					// We don't want to override the ctx passed into Receive, which is necessary for user-initiated cancellations.
+					// This context is passed into the callback (regardless of if tracing is enabled). Thus, it shadows ctx2 so
+					// the callback is cancellable.
+					otelCtx := ctx2
```

Comment:
Should this constructed via context.WithCancel or similar?

Reply:
I'm not sure when we would need to call `cancel` on this. Are you suggesting that we should add the cancel and then check that `otelCtx.Done()` is called?

Reply:
I'm just having trouble following the contexts in use here properly, since you reassign the root ctx2 span to otelCtx each iteration.  For the cases where you don't create  a new span, do any of the usages taint the ctx2 in any way?  There's also https://pkg.go.dev/context#WithoutCancel but I'm not sure that's any better here.

Reply:
> For the cases where you don't create a new span, do any of the usages taint the ctx2 in any way?

No, this doesn't. The assignment of `otelCtx := ctx2` is to create a separate copy of the original context per message (aka per iteration). Modifying `otelCtx` in these cases does not affect the original `ctx2` in any way. This can be demonstrated via a toy example: https://go.dev/play/p/OgTdAcWyy_h

We don't want to use `WithoutCancel` since we want to pass the cancellation signal for ctx2's cancel into the user callback. This is the existing behavior.